src/use-swr-pages.tsx

```diff
@@ -87,68 +87,64 @@ function App () {
   </div>
 }
 */
-
 const pageCacheMap = new Map()
-
 export function useSWRPages<OffsetType = any, Data = any, Error = any>(
   pageKey: string,
   pageFn: pageComponentType<OffsetType, Data, Error>,
   SWRToOffset: pageOffsetMapperType<OffsetType, Data, Error>,
   deps: any[] = []
 ): pagesResponseInterface {
-  const pageCountKey = `_swr_page_count_` + pageKey
-  const pageOffsetKey = `_swr_page_offset_` + pageKey
-
-  const [pageCount, setPageCount] = useState<number>(
-    cacheGet(pageCountKey) || 1
-  )
-  const [pageOffsets, setPageOffsets] = useState<OffsetType[]>(
-    cacheGet(pageOffsetKey) || [null]
-  )
+  const pageCountKey = useMemo(() => `_swr_page_count_` + pageKey, [pageKey])
+  const pageOffsetKey = useMemo(() => `_swr_page_offset_` + pageKey, [pageKey])
+  const [pageCount, setPageCount] = useState(cacheGet(pageCountKey) || 1)
+  const [pageOffsets, setPageOffsets] = useState(cacheGet(pageOffsetKey) || [null])
   const [pageSWRs, setPageSWRs] = useState<responseInterface<Data, Error>[]>([])
-
   const pageFnRef = useRef(pageFn)
   const emptyPageRef = useRef(false)
-
   // Page component (wraps `pageFn`)
   // for performance reason we need to memorize it
   const Page = useCallback(props => {
     // render the page component
     const dataList = pageFnRef.current(props)
-
     // if dataList is [], we can assume this page is empty
     // TODO: this API is not stable
-    if (dataList && !dataList.length) {
-      emptyPageRef.current = true
-    } else {
-      emptyPageRef.current = false
-    }
-
+    emptyPageRef.current = dataList && !dataList.length
     return dataList
   }, [])
+  const resetPages = useCallback(() => {
+    setPageSWRs([])
+    setPageCount(() => {
+      cacheSet(pageCountKey, 1)
+      return 1
+    })
+    setPageOffsets(() => {
+      cacheSet(pageOffsetKey, [null])
+      return [null]
+    })
+  }, [pageOffsetKey, pageCountKey])
 
+  // Reset page if the key changes
+  useEffect(() => resetPages(), [pageOffsetKey])
   // Doesn't have a next page
   const isReachingEnd = pageOffsets[pageCount] === null
-  const isLoadingMore = pageCount === pageOffsets.length
+  const isLoadingMore = pageCount > pageOffsets.length
   const isEmpty = isReachingEnd && pageCount === 1 && emptyPageRef.current
   const loadMore = useCallback(() => {
-    if (isLoadingMore || isReachingEnd) return
+    if (isLoadingMore || isReachingEnd)
+      return
     setPageCount(c => {
       cacheSet(pageCountKey, c + 1)
       return c + 1
     })
   }, [isLoadingMore || isReachingEnd])
```

Comment:
I think we need to add `pageCountKey` as @innocentiv has done here https://github.com/zeit/swr/pull/269/files#diff-c9d2eb3ea821da85bd697b22fe31edf4R153.